streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java

```diff
@@ -302,40 +315,44 @@ protected void flushState() {
     }
 
     private void commitOffsets(final boolean startNewTransaction) {
-        if (commitOffsetNeeded) {
-            log.trace("Committing offsets");
-            final Map<TopicPartition, OffsetAndMetadata> consumedOffsetsAndMetadata = new HashMap<>(consumedOffsets.size());
-            for (final Map.Entry<TopicPartition, Long> entry : consumedOffsets.entrySet()) {
-                final TopicPartition partition = entry.getKey();
-                final long offset = entry.getValue() + 1;
-                consumedOffsetsAndMetadata.put(partition, new OffsetAndMetadata(offset));
-                stateMgr.putOffsetLimit(partition, offset);
-            }
+        try {
+            if (commitOffsetNeeded) {
+                log.trace("Committing offsets");
+                final Map<TopicPartition, OffsetAndMetadata> consumedOffsetsAndMetadata = new HashMap<>(consumedOffsets.size());
+                for (final Map.Entry<TopicPartition, Long> entry : consumedOffsets.entrySet()) {
+                    final TopicPartition partition = entry.getKey();
+                    final long offset = entry.getValue() + 1;
+                    consumedOffsetsAndMetadata.put(partition, new OffsetAndMetadata(offset));
+                    stateMgr.putOffsetLimit(partition, offset);
+                }
 
-            if (eosEnabled) {
-                producer.sendOffsetsToTransaction(consumedOffsetsAndMetadata, applicationId);
+                if (eosEnabled) {
+                    producer.sendOffsetsToTransaction(consumedOffsetsAndMetadata, applicationId);
+                    producer.commitTransaction();
+                    transactionInFlight = false;
+                    if (startNewTransaction) {
+                        transactionInFlight = true;
+                        producer.beginTransaction();
+                    }
+                } else {
+                    try {
+                        consumer.commitSync(consumedOffsetsAndMetadata);
+                    } catch (final CommitFailedException e) {
+                        log.warn("Failed offset commits {} due to CommitFailedException", consumedOffsetsAndMetadata);
+                        throw e;
```

Comment:
Should we move this catch block to the same place as ProducerFenced, and wrap as TaskMigratedException instead? See below.